code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,477 @@
+	var/filled = FALSE
+
+/obj/item/crystal_stabilizer/attackby(obj/item/W, mob/living/user, params)
+	if(!istype(W) || (W.item_flags & ABSTRACT) || !istype(user))
+		return
+	if(istype(W, /obj/item/assembly/signaler/anomaly))
+		if(filled)
+			return
+		to_chat(user, "<span class='notice'>You refill the [src]</span>")
+		filled = TRUE
+		qdel(W)
+
+/*
+This section is for the crystal portals variations
+*/
+/obj/structure/crystal_portal
+	name = "crystal portal"
+	desc = "this shouldn't be here"
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "anom"
+	color = "#B2FFFE"
+	anchored = TRUE
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF | FREEZE_PROOF
+	///Max amount of mobs that a portal can spawn in any given time
+	var/max_mobs = 5
+	///Spawn time between each mobs
+	var/spawn_time = 1000
+	///Type of mob that the portal will spawn, if more than one type in the list will choose randomly
+	var/mob_types = list(/mob/living/simple_animal/hostile/carp)
+	///Fluff text for each mob spawned
+	var/spawn_text = "emerges from"
+	///Affiliation to a faction, used to stop mobs from destroying their portals (unused for now)
+	var/faction = list("hostile")
+	///Spawner component
+	var/spawner_type = /datum/component/spawner
+	///This var check if the portal has been closed by a player with a neutralizer
+	var/closed = FALSE
+
+/obj/structure/crystal_portal/Initialize()
+	. = ..()
+	AddComponent(spawner_type, mob_types, spawn_time, faction, spawn_text, max_mobs)
+	GLOB.crystal_portals += src
+
+/obj/structure/crystal_portal/Destroy()
+	GLOB.crystal_portals -= src
+	if(!closed)
+		switch(name)
+			if("Small Portal")
+				explosion(loc, 0,1,3)
+			if("Medium Portal")
+				explosion(loc, 0,3,5)
+			if("Big Portal")
+				explosion(loc, 1,3,5)
+			if("Huge Portal")
+				explosion(loc, 2,5,7)
+	new/obj/item/assembly/signaler/anomaly(loc)
+	return ..()
+
+/obj/structure/crystal_portal/attack_animal(mob/living/simple_animal/M)
+	if(faction_check(faction, M.faction, FALSE)&&!M.client)
+		return
+	..()
+
+/obj/structure/crystal_portal/attackby(obj/item/W, mob/living/user, params)
+	if(!istype(W) || (W.item_flags & ABSTRACT) || !istype(user))
+		return
+	if(istype(W, /obj/item/anomaly_neutralizer))
+		to_chat(user, "<span class='notice'>You start closing the [src]</span>")
+		if(W.use_tool(src, user, 5.5 SECONDS, volume=100))
+			to_chat(user, "<span class='notice'>You you successfully close the [src]</span>")
+			closed = TRUE
+			qdel(src)
+
+/obj/structure/crystal_portal/small
+	name = "Small Portal"
+	desc = "A small portal to an unkown dimension!"
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "anom"
+	color = "#B2FFFE"
+	anchored = TRUE
+	max_mobs = 3
+	spawn_time = 200
+	mob_types = list(/mob/living/simple_animal/hostile/crystal_monster/minion, /mob/living/simple_animal/hostile/crystal_monster/thug)
+	spawn_text = "emerges from"
+	faction = list("crystal")
+/obj/structure/crystal_portal/medium
```

Comment:
please add some lines between the datums, also, remove the redundant variables.
For example, ``spawn_text`` or ``icon`` are the same for all of these so there is no need to set them for the child types.

Reply:
sure